File: src/main/java/dev/klomptech/jbark/parser/expression/ValueParser.java
package dev.klomptech.jbark.parser.expression;

import dev.klomptech.jbark.config.ConfigLoader;
import dev.klomptech.jbark.errors.BarkError;
import dev.klomptech.jbark.lexer.Token;
import dev.klomptech.jbark.lexer.TokenType;
import dev.klomptech.jbark.parser.Keywords;
import dev.klomptech.jbark.parser.ParseExpression;
import dev.klomptech.jbark.parser.Parser;
import dev.klomptech.jbark.parser.collection.StashSpots;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

// Reads a value from part of the current line (math, trick calls, stash, etc.)
public class ValueParser {

  private final Parser parser;

  // parsed value, and the next word offset to read from
  private record Step(ParseExpression value, int next) {}

  public ValueParser(final Parser parser) {
    this.parser = parser;
  }

  public ParseExpression parsePart(final int startOffset, final int endOffset) throws BarkError {
    if (startOffset >= endOffset) {
      return new ParseExpression.Empty();
    }
    return readExpression(startOffset, endOffset).value();
  }

  // Comma-separated print values, each piece between commas is one expression: growl 2, "hi!"
  public List<ParseExpression> parseCommaSeparated(final int startOffset, final int endOffset)
      throws BarkError {
    List<ParseExpression> values = new ArrayList<>();
    int pos = startOffset;
    while (pos < endOffset) {
      while (pos < endOffset && parser.peekAt(pos).is(TokenType.COMMA)) {
        pos++;
      }
      if (pos >= endOffset) {
        break;
      }
      Step step = readExpression(pos, endOffset);
      values.add(step.value());
      pos = step.next();
    }
    return values;
  }

  // One value on the line: a string, a number, a name, or math like a plus b
  private Step readExpression(final int from, final int to) throws BarkError {
    Step step = readMultiplyOrDivide(from, to);
    while (step.next() < to) {
      ParseExpression.BinaryOp op = addOpAt(step.next());
      if (op == null) {
        break;
      }
      Step right = readMultiplyOrDivide(step.next() + 1, to);
      step = new Step(new ParseExpression.Binary(step.value(), op, right.value()), right.next());
    }
    return step;
  }

  // Example: 2 times 3
  private Step readMultiplyOrDivide(final int from, final int to) throws BarkError {
    Step step = readSingleValue(from, to);
    while (step.next() < to) {
      ParseExpression.BinaryOp op = mulOpAt(step.next());
      if (op == null) {
        break;
      }
      Step right = readSingleValue(step.next() + 1, to);
      step = new Step(new ParseExpression.Binary(step.value(), op, right.value()), right.next());
    }
    return step;
  }

  // One whole value: not false, a number, a string, a trick call, and so on
  private Step readSingleValue(final int from, final int to) throws BarkError {
    if (from >= to) {
      return new Step(new ParseExpression.Empty(), from);
    }
    if (wordAt(from, Keywords.NOT_KEYWORDS)) {
      Step inner = readSingleValue(from + 1, to);
      return new Step(new ParseExpression.Not(inner.value()), inner.next());
    }
    Step start = readNumberStringOrWord(from, to);
    return readWordsAfterValue(start, to);
  }

  // The first piece: a quoted string, a number, or a word
  private Step readNumberStringOrWord(final int from, final int to) throws BarkError {
    Token token = parser.peekAt(from);
    return switch (token.type()) {
      case STRING -> new Step(new ParseExpression.StringLiteral(token.value()), from + 1);
      case NUMBER ->
          new Step(
              new ParseExpression.NumberLiteral(
                  Double.parseDouble(parser.normalise(token.value()))),
              from + 1);
      case IDENTIFIER -> readWord(from, to);
      default ->
          throw new BarkError(
              token.line(), "The dogs don't know what to do with \"" + token.value() + "\" here.");
    };
  }

  // Extra word after a string, like letters in bark "woof" letters
  private Step readWordsAfterValue(final Step step, final int to) {
    int pos = step.next();
    if (pos < to && wordAt(pos, Keywords.LETTERS_WORDS)) {
      return new Step(new ParseExpression.Length(step.value()), pos + 1);
    }
    return step;
  }

  // One identifier: a jar/pile name, nothing, true/false, a trick call, or a plain name
  private Step readWord(final int from, final int to) throws BarkError {
    // Example: "how many toys she has"
    Step howMany = tryParseHowManyFieldCount(from, to);
    if (howMany != null) {
      return howMany;
    }
    // "count of the items in the cookie jar" means how many things are in that jar
    Step jarCount = tryParseItemsInJarCount(from, to);
    if (jarCount != null) {
      return jarCount;
    }
    Step collection = readCollection(from, to);
    if (collection != null) {
      return collection;
    }
    String word = parser.normalise(parser.peekAt(from).value());
    if (Keywords.TYPE_NOTHING_WORDS.contains(word) || Keywords.NULL_VALUES.contains(word)) {
      return new Step(new ParseExpression.NullLiteral(), from + 1);
    }
    Optional<Boolean> bool = Keywords.parseBoolean(word);
    if (bool.isPresent()) {
      return new Step(new ParseExpression.BooleanLiteral(bool.get()), from + 1);
    }
    if (from + 1 < to && wordAt(from + 1, Keywords.WITH_WORDS)) {
      return readTrickCall(word, from + 2, to);
    }
    return new Step(new ParseExpression.Variable(word), from + 1);
  }

  // Example: "how many treats Bimba has" or "how many toys the labrador has"
  private Step tryParseHowManyFieldCount(final int from, final int to) {
    if (from + 2 >= to
        || !wordAt(from, Keywords.HOW_WORDS)
        || !wordAt(from + 1, Keywords.MANY_WORDS)) {
      return null;
    }
    int topicOffset = from + 2;
    if (!parser.peekAt(topicOffset).is(TokenType.IDENTIFIER)) {
      return null;
    }
    String topic =
        Keywords.explicitAttributeTopic(parser.normalise(parser.peekAt(topicOffset).value()));
    if (topic == null) {
      return null;
    }
    int whoStart = topicOffset + 1;
    if (whoStart >= to) {
      return null;
    }
    String who =
        ConfigLoader.resolveNameFromTokens(
            parser.allTokens(), parser.cursorIndex() + whoStart, parser.cursorIndex() + to);
    int next;
    if (who == null) {
      Token whoToken = parser.peekAt(whoStart);
      if (!whoToken.is(TokenType.IDENTIFIER)) {
        return null;
      }
      String whoWord = parser.normalise(whoToken.value());
      if (!Keywords.isDogSubjectReference(whoWord, whoToken.value())) {
        return null;
      }
      who = whoWord;
      next = whoStart + 1;
    } else {
      int whoEnd =
          ConfigLoader.tokenIndexAfterResolvedName(
              parser.allTokens(), parser.cursorIndex() + whoStart, parser.cursorIndex() + to, who);
      next = whoEnd < 0 ? whoStart + 1 : whoEnd - parser.cursorIndex();
    }
    if (next < to && Keywords.isAssignSubjectGlue(parser.normalise(parser.peekAt(next).value()))) {
      next++;
    }
    return new Step(new ParseExpression.Field(who, topic), next);
  }

  // Phrase: count of items in <jar or pile>. Story words between those keywords are ignored.
  private Step tryParseItemsInJarCount(final int from, final int to) throws BarkError {
    if (!wordAt(from, Keywords.COUNT_WORDS)) {
      return null;
    }
    int pos = from + 1;
    if (pos >= to || !wordAt(pos, Keywords.OF_WORDS)) {
      return null;
    }
    pos = findNextPhraseWord(pos + 1, to, Keywords.ITEMS_WORDS);
    if (pos < 0) {
      return null;
    }
    pos = findNextPhraseWord(pos + 1, to, Keywords.IN_WORDS);
    if (pos < 0) {
      return null;
    }
    int afterIn = pos + 1;
    String name =
        ConfigLoader.resolveCollectionFromTokens(
            parser.allTokens(), parser.cursorIndex() + afterIn, parser.cursorIndex() + to);
    if (name == null) {
      throw new BarkError(
          parser.peekAt(from).line(),
          "The dogs were waiting for a jar to investigate, but you never gave them one!");
    }
    int afterName = afterIn;
    for (int end = afterIn + 1; end <= Math.min(to, afterIn + 4); end++) {
      String found =
          ConfigLoader.resolveCollectionFromTokens(
              parser.allTokens(), parser.cursorIndex() + afterIn, parser.cursorIndex() + end);
      if (name.equals(found)) {
        afterName = end;
        break;
      }
    }
    if (ConfigLoader.isStash(name)) {
      return new Step(
          new ParseExpression.StashAccess(name, ParseExpression.StashPart.COUNT, null), afterName);
    }
    if (ConfigLoader.isPile(name)) {
      return new Step(
          new ParseExpression.PileAccess(name, ParseExpression.PilePart.COUNT), afterName);
    }
    // A name followed "in" but it is not a registered jar or pile
    throw new BarkError(
        parser.peekAt(from).line(),
        "The dogs are confused by what you wanted them to investigate! That was not a jar!");
  }

  // Scan forward for the next required phrase word; skip anything the dogs do not recognize here
  private int findNextPhraseWord(final int from, final int to, final List<String> words) {
    for (int pos = from; pos < to; pos++) {
      if (wordAt(pos, words)) {
        return pos;
      }
    }
    return -1;
  }

  // cookie_jar or her cookie jar when the line is talking about a jar or pile
  private Step readCollection(final int from, final int to) throws BarkError {
    int absStart = parser.cursorIndex() + from;
    int absEnd = parser.cursorIndex() + to;
    String name = ConfigLoader.resolveCollectionFromTokens(parser.allTokens(), absStart, absEnd);
    if (name == null) {
      return null;
    }
    int afterName = countTokensInCollectionName(from, to, name);
    if (ConfigLoader.isStash(name)) {
      return readStashValue(name, afterName, to);
    }
    if (ConfigLoader.isPile(name)) {
      return readPileValue(name, afterName, to);
    }
    return null;
  }

  // How many words the jar or pile name takes (cookie_jar is one word, her cookie jar is three)
  private int countTokensInCollectionName(final int from, final int to, final String name) {
    for (int end = from + 1; end <= Math.min(to, from + 4); end++) {
      String found =
          ConfigLoader.resolveCollectionFromTokens(
              parser.allTokens(), parser.cursorIndex() + from, parser.cursorIndex() + end);
      if (name.equals(found)) {
        return end;
      }
    }
    return from + 1;
  }

  // After the jar name: all items, one slot, or count
  private Step readStashValue(final String stash, final int from, final int to) throws BarkError {
    int pos = parser.skipIgnoredIdentifiers(from, to);
    if (pos < to && parser.peekAt(pos).is(TokenType.IDENTIFIER)) {
      String spot = parser.normalise(parser.peekAt(pos).value());
      if (StashSpots.isStashSpot(spot)) {
        return new Step(
            new ParseExpression.StashAccess(stash, ParseExpression.StashPart.ELEMENT, spot),
            pos + 1);
      }
    }
    if (pos < to && wordAt(pos, Keywords.COUNT_WORDS)) {
      return new Step(
          new ParseExpression.StashAccess(stash, ParseExpression.StashPart.COUNT, null), pos + 1);
    }
    return new Step(
        new ParseExpression.StashAccess(stash, ParseExpression.StashPart.ALL, null), pos);
  }

  // bark laundry_basket prints everything in that pile
  private Step readPileValue(final String pile, final int from, final int to) {
    int pos = parser.skipIgnoredIdentifiers(from, to);
    return new Step(new ParseExpression.PileAccess(pile, ParseExpression.PilePart.ALL), pos);
  }

  // Example: add with 2, 3
  private Step readTrickCall(final String name, final int from, final int to) throws BarkError {
    List<ParseExpression> args = new ArrayList<>();
    int pos = from;
    while (pos < to) {
      Step arg = readExpression(pos, to);
      args.add(arg.value());
      pos = arg.next();
      if (pos < to && Keywords.isListSeparator(parser.peekAt(pos))) {
        pos++;
        continue;
      }
      break;
    }
    return new Step(new ParseExpression.FunctionCall(name, args), pos);
  }

  // Is this word plus or minus?
  private ParseExpression.BinaryOp addOpAt(final int offset) {
    Token token = parser.peekAt(offset);
    if (token.is(TokenType.PLUS)) {
      return ParseExpression.BinaryOp.PLUS;
    }
    if (token.is(TokenType.MINUS)) {
      return ParseExpression.BinaryOp.MINUS;
    }
    if (token.is(TokenType.IDENTIFIER)) {
      String word = parser.normalise(token.value());
      if (Keywords.isMathAddKeyword(word)) {
        return ParseExpression.BinaryOp.PLUS;
      }
      if (Keywords.isMathSubtractKeyword(word)) {
        return ParseExpression.BinaryOp.MINUS;
      }
    }
    return null;
  }

  // Is this word times or divided by?
  private ParseExpression.BinaryOp mulOpAt(final int offset) {
    Token token = parser.peekAt(offset);
    if (token.is(TokenType.STAR)) {
      return ParseExpression.BinaryOp.STAR;
    }
    if (token.is(TokenType.SLASH)) {
      return ParseExpression.BinaryOp.SLASH;
    }
    if (token.is(TokenType.IDENTIFIER)) {
      String word = parser.normalise(token.value());
      if (Keywords.STAR_KEYWORDS.contains(word)) {
        return ParseExpression.BinaryOp.STAR;
      }
      if (Keywords.SLASH_KEYWORDS.contains(word)) {
        return ParseExpression.BinaryOp.SLASH;
      }
    }
    return null;
  }

  private boolean wordAt(final int offset, final List<String> words) {
    Token token = parser.peekAt(offset);
    return token.is(TokenType.IDENTIFIER) && words.contains(parser.normalise(token.value()));
  }
}

File: pbark/pbark/parser/expression/value_parser.py
from __future__ import annotations

from dataclasses import dataclass

from pbark.config.config_loader import ConfigLoader
from pbark.errors import BarkError
from pbark.lexer import TokenType
from pbark.parser import keywords as Keywords
from pbark.parser.collection import stash_spots as StashSpots
from pbark.parser.parse_expression import (
    Binary,
    BinaryOp,
    BooleanLiteral,
    Empty,
    Field,
    FunctionCall,
    Length,
    NullLiteral,
    NumberLiteral,
    ParseExpression,
    PileAccess,
    PilePart,
    StashAccess,
    StashPart,
    StringLiteral,
    Variable,
    Not,
)


@dataclass
class _Step:
    value: ParseExpression
    next: int


class ValueParser:
    def __init__(self, parser) -> None:
        self._parser = parser

    def parse_part(self, start_offset: int, end_offset: int) -> ParseExpression:
        if start_offset >= end_offset:
            return Empty()
        return self._read_expression(start_offset, end_offset).value

    def parse_comma_separated(self, start_offset: int, end_offset: int) -> list[ParseExpression]:
        values: list[ParseExpression] = []
        pos = start_offset
        while pos < end_offset:
            while pos < end_offset and self._parser.peek_at(pos).is_(TokenType.COMMA):
                pos += 1
            if pos >= end_offset:
                break
            step = self._read_expression(pos, end_offset)
            values.append(step.value)
            pos = step.next
        return values

    def _read_expression(self, from_: int, to: int) -> _Step:
        step = self._read_multiply_or_divide(from_, to)
        while step.next < to:
            op = self._add_op_at(step.next)
            if op is None:
                break
            right = self._read_multiply_or_divide(step.next + 1, to)
            step = _Step(Binary(step.value, op, right.value), right.next)
        return step

    def _read_multiply_or_divide(self, from_: int, to: int) -> _Step:
        step = self._read_single_value(from_, to)
        while step.next < to:
            op = self._mul_op_at(step.next)
            if op is None:
                break
            right = self._read_single_value(step.next + 1, to)
            step = _Step(Binary(step.value, op, right.value), right.next)
        return step

    def _read_single_value(self, from_: int, to: int) -> _Step:
        if from_ >= to:
            return _Step(Empty(), from_)
        if self._word_at(from_, Keywords.NOT_KEYWORDS):
            inner = self._read_single_value(from_ + 1, to)
            return _Step(Not(inner.value), inner.next)
        start = self._read_number_string_or_word(from_, to)
        return self._read_words_after_value(start, to)

    def _read_number_string_or_word(self, from_: int, to: int) -> _Step:
        token = self._parser.peek_at(from_)
        if token.is_(TokenType.STRING):
            return _Step(StringLiteral(token.value), from_ + 1)
        if token.is_(TokenType.NUMBER):
            return _Step(
                NumberLiteral(float(self._parser.normalise(token.value))),
                from_ + 1,
            )
        if token.is_(TokenType.IDENTIFIER):
            return self._read_word(from_, to)
        raise BarkError(
            token.line,
            f'The dogs don\'t know what to do with "{token.value}" here.',
        )

    def _read_words_after_value(self, step: _Step, to: int) -> _Step:
        pos = step.next
        if pos < to and self._word_at(pos, Keywords.LETTERS_WORDS):
            return _Step(Length(step.value), pos + 1)
        return step

    def _read_word(self, from_: int, to: int) -> _Step:
        how_many = self._try_parse_how_many_field_count(from_, to)
        if how_many is not None:
            return how_many
        jar_count = self._try_parse_items_in_jar_count(from_, to)
        if jar_count is not None:
            return jar_count
        collection = self._read_collection(from_, to)
        if collection is not None:
            return collection
        word = self._parser.normalise(self._parser.peek_at(from_).value)
        if word in Keywords.TYPE_NOTHING_WORDS or word in Keywords.NULL_VALUES:
            return _Step(NullLiteral(), from_ + 1)
        bool_val = Keywords.parse_boolean(word)
        if bool_val is not None:
            return _Step(BooleanLiteral(bool_val), from_ + 1)
        if from_ + 1 < to and self._word_at(from_ + 1, Keywords.WITH_WORDS):
            return self._read_trick_call(word, from_ + 2, to)
        return _Step(Variable(word), from_ + 1)

    def _try_parse_how_many_field_count(self, from_: int, to: int) -> _Step | None:
        if (
            from_ + 2 >= to
            or not self._word_at(from_, Keywords.HOW_WORDS)
            or not self._word_at(from_ + 1, Keywords.MANY_WORDS)
        ):
            return None
        topic_offset = from_ + 2
        if not self._parser.peek_at(topic_offset).is_(TokenType.IDENTIFIER):
            return None
        topic = Keywords.explicit_attribute_topic(
            self._parser.normalise(self._parser.peek_at(topic_offset).value)
        )
        if topic is None:
            return None
        who_start = topic_offset + 1
        if who_start >= to:
            return None
        who = ConfigLoader.resolve_name_from_tokens(
            self._parser.all_tokens(),
            self._parser.cursor_index() + who_start,
            self._parser.cursor_index() + to,
        )
        next_pos: int
        if who is None:
            who_token = self._parser.peek_at(who_start)
            if not who_token.is_(TokenType.IDENTIFIER):
                return None
            who_word = self._parser.normalise(who_token.value)
            if not Keywords.is_dog_subject_reference(who_word, who_token.value):
                return None
            who = who_word
            next_pos = who_start + 1
        else:
            who_end = ConfigLoader.token_index_after_resolved_name(
                self._parser.all_tokens(),
                self._parser.cursor_index() + who_start,
                self._parser.cursor_index() + to,
                who,
            )
            next_pos = who_start + 1 if who_end < 0 else who_end - self._parser.cursor_index()
        if next_pos < to and Keywords.is_assign_subject_glue(
            self._parser.normalise(self._parser.peek_at(next_pos).value)
        ):
            next_pos += 1
        return _Step(Field(who, topic), next_pos)

    def _try_parse_items_in_jar_count(self, from_: int, to: int) -> _Step | None:
        if not self._word_at(from_, Keywords.COUNT_WORDS):
            return None
        pos = from_ + 1
        if pos >= to or not self._word_at(pos, Keywords.OF_WORDS):
            return None
        pos = self._find_next_phrase_word(pos + 1, to, Keywords.ITEMS_WORDS)
        if pos < 0:
            return None
        pos = self._find_next_phrase_word(pos + 1, to, Keywords.IN_WORDS)
        if pos < 0:
            return None
        after_in = pos + 1
        name = ConfigLoader.resolve_collection_from_tokens(
            self._parser.all_tokens(),
            self._parser.cursor_index() + after_in,
            self._parser.cursor_index() + to,
        )
        if name is None:
            raise BarkError(
                self._parser.peek_at(from_).line,
                "The dogs were waiting for a jar to investigate, but you never gave them one!",
            )
        after_name = after_in
        for end in range(after_in + 1, min(to, after_in + 4) + 1):
            found = ConfigLoader.resolve_collection_from_tokens(
                self._parser.all_tokens(),
                self._parser.cursor_index() + after_in,
                self._parser.cursor_index() + end,
            )
            if name == found:
                after_name = end
                break
        if ConfigLoader.is_stash(name):
            return _Step(StashAccess(name, StashPart.COUNT), after_name)
        if ConfigLoader.is_pile(name):
            return _Step(PileAccess(name, PilePart.COUNT), after_name)
        raise BarkError(
            self._parser.peek_at(from_).line,
            "The dogs are confused by what you wanted them to investigate! That was not a jar!",
        )

    def _find_next_phrase_word(self, from_: int, to: int, words: list[str]) -> int:
        for pos in range(from_, to):
            if self._word_at(pos, words):
                return pos
        return -1

    def _read_collection(self, from_: int, to: int) -> _Step | None:
        abs_start = self._parser.cursor_index() + from_
        abs_end = self._parser.cursor_index() + to
        name = ConfigLoader.resolve_collection_from_tokens(
            self._parser.all_tokens(), abs_start, abs_end
        )
        if name is None:
            return None
        after_name = self._count_tokens_in_collection_name(from_, to, name)
        if ConfigLoader.is_stash(name):
            return self._read_stash_value(name, after_name, to)
        if ConfigLoader.is_pile(name):
            return self._read_pile_value(name, after_name, to)
        return None

    def _count_tokens_in_collection_name(self, from_: int, to: int, name: str) -> int:
        for end in range(from_ + 1, min(to, from_ + 4) + 1):
            found = ConfigLoader.resolve_collection_from_tokens(
                self._parser.all_tokens(),
                self._parser.cursor_index() + from_,
                self._parser.cursor_index() + end,
            )
            if name == found:
                return end
        return from_ + 1

    def _read_stash_value(self, stash: str, from_: int, to: int) -> _Step:
        pos = self._parser.skip_ignored_identifiers(from_, to)
        if pos < to and self._parser.peek_at(pos).is_(TokenType.IDENTIFIER):
            spot = self._parser.normalise(self._parser.peek_at(pos).value)
            if StashSpots.is_stash_spot(spot):
                return _Step(StashAccess(stash, StashPart.ELEMENT, spot), pos + 1)
        if pos < to and self._word_at(pos, Keywords.COUNT_WORDS):
            return _Step(StashAccess(stash, StashPart.COUNT), pos + 1)
        return _Step(StashAccess(stash, StashPart.ALL), pos)

    def _read_pile_value(self, pile: str, from_: int, to: int) -> _Step:
        pos = self._parser.skip_ignored_identifiers(from_, to)
        return _Step(PileAccess(pile, PilePart.ALL), pos)

    def _read_trick_call(self, name: str, from_: int, to: int) -> _Step:
        args: list[ParseExpression] = []
        pos = from_
        while pos < to:
            arg = self._read_expression(pos, to)
            args.append(arg.value)
            pos = arg.next
            if pos < to and Keywords.is_list_separator(self._parser.peek_at(pos)):
                pos += 1
                continue
            break
        return _Step(FunctionCall(name, args), pos)

    def _add_op_at(self, offset: int) -> BinaryOp | None:
        token = self._parser.peek_at(offset)
        if token.is_(TokenType.PLUS):
            return BinaryOp.PLUS
        if token.is_(TokenType.MINUS):
            return BinaryOp.MINUS
        if token.is_(TokenType.IDENTIFIER):
            word = self._parser.normalise(token.value)
            if Keywords.is_math_add_keyword(word):
                return BinaryOp.PLUS
            if Keywords.is_math_subtract_keyword(word):
                return BinaryOp.MINUS
        return None

    def _mul_op_at(self, offset: int) -> BinaryOp | None:
        token = self._parser.peek_at(offset)
        if token.is_(TokenType.STAR):
            return BinaryOp.STAR
        if token.is_(TokenType.SLASH):
            return BinaryOp.SLASH
        if token.is_(TokenType.IDENTIFIER):
            word = self._parser.normalise(token.value)
            if word in Keywords.STAR_KEYWORDS:
                return BinaryOp.STAR
            if word in Keywords.SLASH_KEYWORDS:
                return BinaryOp.SLASH
        return None

    def _word_at(self, offset: int, words: list[str]) -> bool:
        token = self._parser.peek_at(offset)
        return token.is_(TokenType.IDENTIFIER) and self._parser.normalise(token.value) in words
